### PR TITLE
Improve resources disposal

### PR DESCRIPTION
**Pull request by jailln**
## Description

Improve resource disposal by:
* using `ObjectRemovalHelper` to remove 3D objects when a `GeometryLayer` is deleted
* disposing textures in `ObjectRemovalHelper`
* allowing to clear the cache or not when disposing the view and layers

## Motivation and Context

Some resources where not disposed after disposing itowns view. To make sure of it, you can add `setTimeout(function() {view.dispose();}, 40000);` at the end of any itowns example. Some tests I made with the chrome Memory tool, on two itowns examples:

### Before the MR

**3dtiles_basic.html:**

*Before disposing the view*
3dtiles_basic_before

*After disposing the view*
3dtiles_basic_after

Notice that nearly any memory is freed and that there is still `ArrayBuffers` which contains geometries

**vector_tile_3d_mesh.html:**

*Before disposing the view*
vector_tile_3d_mesh_before

*After disposing the view*
vector_tile_3d_mesh_after

Same here, not much is deleted

### After the MR

*Before disposing the view*
image

*After disposing the view*
image

It's not perfect but more resources are disposed.

**vector_tile_3d_mesh.html:**

I have less success with this one :grin: 

*Before disposing the view*
image

*After disposing the view*
image

There is still many `Feature` and `FeatureGeometry` after the view disposal, but I haven't found out where they are stored and how to delete them. Note that the cache and the threejs object3Ds are deleted the same way than for `c3DTileslayer` since both `c3DTilesLayer` and `FeatureGeometryLayer` inherit from `GeometryLayer`.  Help wanted for this one :slightly_smiling_face: 

## Open discussion

I think we could still improve some other things. Let me know what you think, if you agree I will implement them:
* Harmonize functions names for resource disposal (some are named dispose, other delete, other clear, other remove, etc.): rename them all to `dispose` (the name used in threejs)
* Some caches are declared as constants outside classes and are not exported: `cacheStyle` in `Style.js`, `cacheTile` and `cacheBuffer` in `TileBuilder.js`. Therefore they cannot be cleared properly when disposing the view. I think they should be stored in the view so we can clear them when needed.
* Fetched data are cached twice:
  * Once after being parsed, in `Source._featuresCaches`
  * Once after being converted, in `Layer.cache`
I think this greatly affects itowns memory footprint, especially when loading heavy dataset. In addition, I don't think there is a case where only one cache is deleted: they are both deleted when the layer is deleted (or when the cache lifetime has expired). WDYT about caching them only once ? (in `Layer.cache` to avoid transforming them again)

### CODE REVIEW

File: src/Process/LayeredMaterialNodeProcessing.js [hunk 1]
```diff
@@ -237,6 +237,9 @@ export function updateLayeredMaterialNodeElevation(context, layer, node, parent)
 
 export function removeLayeredMaterialNodeLayer(layerId) {
     return function removeLayeredMaterialNodeLayer(node) {
+        if (node.geometry) {
+            node.geometry.dispose();
+        }
```
Comment (mgermerie) [inline src/Process/LayeredMaterialNodeProcessing.js]: I don't think the geometry should be disposed when removing a `RasterLayer`. The geometry of each tile should not depend on layers attached to the tile. This allows us to remove a `ColorLayer` without causing issue for an `ElevationLayer` visualization for instance.

On the screenshots bellow, you can se the effect caused by disposing each tile geometry when removing a ColorLayer, and what should happen (without disposing geometry).

Before removing `ColorLayer` :


After removing `ColorLayer` and disposing tiles geometry : 


What should normaly happen (if you don't dispose geometry) : 


For this third one, this is not exactly what happens, I just ran into some glitch and I shall make an issue about it (some of the tiles do not display elevation data until the camera moves and the view is updated. However, the tiles are still displayed, unlike in the second screenshot).

File: src/Layer/OrientedImageLayer.js [hunk 2]
```diff
@@ -205,9 +205,12 @@ class OrientedImageLayer extends GeometryLayer {
      * Delete background, but doesn't delete OrientedImageLayer.material. For the moment, this material visibility is set to false.
      * You need to replace OrientedImageLayer.material applied on each object, if you want to continue displaying them.
      * This issue (see #1018 {@link https://github.com/iTowns/itowns/issues/1018}) will be fixed when OrientedImageLayer will be a ColorLayer.
-     */
-    delete() {
-        super.delete();
+    * @param {boolean} [clearCache=false] Whether to clear the layer cache or not
+    */
+    delete(clearCache) {
+        if (clearCache) {
```
Comment (mgermerie) [inline src/Layer/OrientedImageLayer.js:219]: The call to `super.delete()` is indeed not needed if `background` property for `OrientedImageLayer` has not been set. In this case, `OrientedImageLayer.object3d` is just an empty `THREE.Group()`, with nothing to dispose.

However, if `background` property is set (as it is the case in [immersive view example](http://www.itowns-project.org/itowns/examples/#view_immersive)), `OrientedImageLayer.object3d` can contain meshes. Therefore, these meshes need to be disposed by calling `super.delete()`.

In the `GeometryLayer.delete` method modified with this PR, the objects that are disposed are filtered (only those with the correct `layer.id` property are disposed, which was not the case before). The issue with `OrientedImageLayer` objects is that they lack the `layer.id` property, which should therefore be added.

To summarise : I think we should add `super.delete()` back, and add something like the two following lines in `OrientedImageLayer` constructor, just above line 115.

```js
this.background.layer = this.background.layer ?? {};
this.background.layer.id = id;
```

This will allow `OrientedImageLayer.background` meshes to be correctly disposed if they exist.

What do you think about this ?

File: src/Process/LayeredMaterialNodeProcessing.js (hunk 1 above)
Comment (jailln) [reply src/Process/LayeredMaterialNodeProcessing.js]: I agree with you that `The geometry of each tile should not depend on layers attached to the tile.` 
However, I thought that `node.geometry` represented the own geometry of the `RasterLayer` if it was not attached, but now I'm not sure of it. Can a`RasterLayer` have its own geometry or is it always attached to a GeometryLayer? Is `node.geometry` always the geometry of the layer it is attached too?

File: src/Layer/OrientedImageLayer.js (hunk 2 above)
Comment (jailln) [reply src/Layer/OrientedImageLayer.js:219]: Yep I think you're right, I will make the modifications :)

Side question: Regarding the filtering of the geometries by layer id in ObjectRemovalHelper, which usecase does it serve? I cannot think of a case where a GeometryLayer would have geometries that don't have the id of this geometry layer, is there one ?

File: src/Process/LayeredMaterialNodeProcessing.js (hunk 1 above)
Comment (mgermerie) [reply src/Process/LayeredMaterialNodeProcessing.js]: `RasterLayers` are always attached to another `GeometryLayer` - which most of the times is a [`TiledGeometryLayer`](https://github.com/iTowns/itowns/blob/master/src/Layer/TiledGeometryLayer.js). This `TiledGeometryLayer` displays the globe -or planar- tiles with a default material (the blue color we see if no `ColorLayer` is attached). When we add a `RasterLayer` : 
- if it is a `ColorLayer`, the image is used as a texture for the tile material,
- if it is an `ElevationLayer`, the `z` coordinate of each tile vertex is translated by the correct elevation.

When removing a `RasterLayer`, we take the tiles from the minimum zoom level, and we traverse all their children which are passed to `removeLayeredMaterialNodeLayer` parameter. Therefore, the `node` in this case is always a tile from the `TiledGeometryLayer` to which the `RasterLayer` is attached.

To conclude and answer your question : 
- a `RasterLayer` does not have its own geometry and is always attached to a `GeometryLayer`.
- `node.geometry` is always the geometry of the layer the `RasterLayer` is attached to.

I hope this is clear enough, don't hesitate if it needs clarification.

_PS : each tile from a `TiledGeometryLayer` actually has five children : the four tiles matching the subsequent zoom level subdivision, and the tile's OBB. Perhap this could be improved, having the OBB as a property and not a children._

**Comment by jailln** [comment]
My remarks regarding the cache have been moved to #1928

File: src/Layer/OrientedImageLayer.js (hunk 2 above)
Comment (mgermerie) [reply src/Layer/OrientedImageLayer.js:219]: In case a `LabelLayer` is added, the labels are added as children of a tile. Sometimes, the method is called with the tile as object parameter. The filtering prevents removing labels.

File: src/Process/LayeredMaterialNodeProcessing.js (hunk 1 above)
Comment (jailln) [reply src/Process/LayeredMaterialNodeProcessing.js]: Ok thanks for the explaination, it is clearer now! I removed this line and I added your explanation in the documentation.

Can you open issues about the problem you mention and about your proposition for OBB management please?

File: src/Layer/OrientedImageLayer.js (hunk 2 above)
Comment (jailln) [reply src/Layer/OrientedImageLayer.js:219]: Done

**Comment by jailln** [comment]
I also created issues #1937 and #1936 for the remaining remarks of my PR description.

File: src/Layer/OrientedImageLayer.js [hunk 3]
```diff
@@ -205,9 +209,16 @@ class OrientedImageLayer extends GeometryLayer {
...
      * This issue (see #1018 {@link https://github.com/iTowns/itowns/issues/1018}) will be fixed when OrientedImageLayer will be a ColorLayer.
-     */
-    delete() {
-        super.delete();
+    * @param {boolean} [clearCache=false] Whether to clear the layer cache or not
+    */
+    delete(clearCache) {
+        if (this.background) {
+            // only delete geometries if it has some
+            super.delete();
```
Comment (mgermerie) [inline src/Layer/OrientedImageLayer.js:217]: I noticed an inconsistency regarding this part : 

An `OrientedImageLayer` is never attached to a `GeometryLayer`. It has an empty `update` method which does not depend on a parent layer (i.e. a layer it could be attached to). The projection is handled in the shader.  Therefore, the `OrientedImageLayer.parent` property should always be `undefined`. 

However, when adding an `OrientedImageLayer` in a `GlobeView`, the `GlobeView.addLayer` method called enforce each layer's parent to its `TiledGeometryLayer` (i.e. the globe terrain). It leaves us with `OrientedImageLayers` whose `parent` property is the `GlobeLayer`.

This does not causes issue I am aware of. I just noticed it while looking if testing `this.background` was redundant with `GeometryLayer.delete` tests (and you're wright, it is not :smiley: )

Comment (jailln) [reply src/Layer/OrientedImageLayer.js:217]: Ahah ok, maybe we should open an issue about this ? And maybe `OrientedImageLayer` should be added `View`'s `addLayer` then ? : `itowns.View.prototype.addLayer.call(view, orientedImgLayer);`

**Comment by jailln** [comment]
@mgermerie could you create the issues of the remaining points you noted and add your review please? I need this MR to solve #1947 :slightly_smiling_face:

File: src/Process/LayeredMaterialNodeProcessing.js (hunk 1 above)
Comment (mgermerie) [reply src/Process/LayeredMaterialNodeProcessing.js]: #1983 opened on this subject.

File: src/Layer/OrientedImageLayer.js (hunk 3 above)
Comment (mgermerie) [reply src/Layer/OrientedImageLayer.js:217]: #1984 opened on the subject.

File: src/Process/ObjectRemovalHelper.js [hunk 4]
```diff
@@ -1,3 +1,14 @@
...
+    material.dispose();
+    // dispose textures
+    for (const key of Object.keys(material)) {
+        const val = material[key];
+        if (val && val.isTexture) {
+            val.dispose();
+        }
+    }
+}
+
```
Comment (gchoqueux) [inline src/Process/ObjectRemovalHelper.js:11]: what type of materials do you want to dispose?

Comment (jailln) [reply src/Process/ObjectRemovalHelper.js:11]: I made it generic so all type of materials and their textures can be disposed with this method

**Comment by jailln** [comment]
commits squashed and branch rebased

**Comment by jailln** [comment]
@gchoqueux @mgermerie is it ok for you? can we merge it ?

**Review by gchoqueux — APPROVED** [review]
it's ok for me
